client/store/cart.js

```diff
@@ -1,35 +1,103 @@
 import axios from "axios";
 
 const SET_CART = "SET_CART";
-const UPDATE_CART = "UPDATE_CART";
+const ADD_TO_CART = "ADD_TO_CART";
+const DECREASE_ITEM = "DECREASE_ITEM";
+const REMOVE_FROM_CART = "REMOVE_FROM_CART";
 
 export const setCart = (cart) => ({
   type: SET_CART,
   cart,
 });
 
-export const _updateCart = (cart) => ({
-  type: UPDATE_CART,
+export const addItem = (cart) => ({
+  type: ADD_TO_CART,
   cart,
 });
 
-export const fetchCart = () => {
+export const _decreaseItem = (cart) => ({
+  type: DECREASE_ITEM,
+  cart,
+});
+
+export const removeItem = (cart) => ({
+  type: REMOVE_FROM_CART,
+  cart,
+});
+
+export const fetchCart = (userId) => {
   return async (dispatch) => {
     try {
-      const { data } = await axios.get("/api/cart");
+      const { data } = await axios.get("/api/cart", { userId });
       dispatch(setCart(data));
     } catch (error) {
       console.log(error);
     }
   };
 };
 
-const initialState = [];
+export const clearCart = () => (dispatch) => {
+  dispatch(setCart([]));
+  window.localStorage.setItem("cart", JSON.stringify([]));
+};
+
+export const addToCart = (product) => (dispatch, getState) => {
+  const cartItems = getState().cart.cartItems.slice();
+  let inCart = false;
+  cartItems.forEach((item) => {
+    if (item.name === product.name) {
+      inCart = true;
+      item.count++;
+    }
+  });
+  if (!inCart) {
+    cartItems.push({ ...product, count: 1 });
+  }
+  dispatch(addItem(cartItems));
+  window.localStorage.setItem("cart", JSON.stringify(cartItems));
+};
```

Comment:
addToCart pulls current cart from state and makes a copy of it in line 45, then we check if the item already exists in the cart, if it does, we simply increase the count, otherwise we push the item to the end of the cart array and set count to 1. we also update the cart in localstorage in order to persist the cart if we are a guest/not logged in